test/e2e/allowed_source_ranges_test.go

```diff
@@ -0,0 +1,396 @@
+// and verifies that it is reflected in AllowedSourceRanges instead of
+// the annotation as the field takes precedence over the annotation.
+func TestAllowedSourceRangesStatus(t *testing.T) {
+	t.Parallel()
+
+	if infraConfig.Status.PlatformStatus == nil {
+		t.Skip("test skipped on nil platform")
+	}
+	supportedPlatforms := map[configv1.PlatformType]struct{}{
+		configv1.AWSPlatformType:   {},
+		configv1.AzurePlatformType: {},
+		configv1.GCPPlatformType:   {},
+	}
+	if _, supported := supportedPlatforms[infraConfig.Status.PlatformStatus.Type]; !supported {
+		t.Skipf("test skipped on platform %q", infraConfig.Status.PlatformStatus.Type)
+	}
+
+	// Create an ingresscontroller with a loadbalancer endpoint publishing strategy.
+	name := types.NamespacedName{Namespace: operatorNamespace, Name: "sourcerangesstatus"}
+	domain := name.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newLoadBalancerController(name, domain)
+	ic.Spec.EndpointPublishingStrategy.LoadBalancer = &operatorv1.LoadBalancerStrategy{
+		Scope:               operatorv1.ExternalLoadBalancer,
+		DNSManagementPolicy: operatorv1.ManagedLoadBalancerDNS,
+	}
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
+	}
+	t.Cleanup(func() { assertIngressControllerDeleted(t, kclient, ic) })
+
+	// Wait for the load balancer and DNS to be ready.
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, name, availableConditionsForIngressControllerWithLoadBalancer...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	lbService := &corev1.Service{}
+	if err := kclient.Get(context.TODO(), controller.LoadBalancerServiceName(ic), lbService); err != nil {
+		t.Fatalf("failed to get service: %v", err)
+	}
+
+	// Set service.beta.kubernetes.io/load-balancer-source-ranges annotation and see if it will be reflected in AllowedSourceRanges in IngressController's status
+	lbService.Annotations[ingresscontroller.LoadBalancerSourceRangesAnnotation] = "127.0.0.0/8"
+	if err := kclient.Update(context.TODO(), lbService); err != nil {
+		t.Fatalf("failed to update service: %v", err)
+	}
+
+	if err := kclient.Get(context.TODO(), name, ic); err != nil {
+		t.Fatalf("failed to get ingresscontroller %s: %v", name, err)
+	}
+	err := wait.PollImmediate(10*time.Second, 2*time.Minute, func() (bool, error) {
+		cidrs := []string{}
+		if ic.Status.EndpointPublishingStrategy != nil {
+			lb := ic.Status.EndpointPublishingStrategy.LoadBalancer
+			if lb != nil && len(lb.AllowedSourceRanges) > 0 {
+				for _, cidr := range lb.AllowedSourceRanges {
+					cidrs = append(cidrs, string(cidr))
+				}
+			}
+		}
+
+		annotationRanges := strings.Split(lbService.Annotations[ingresscontroller.LoadBalancerSourceRangesAnnotation], ",")
+		t.Logf("Annotation: %v, AllowedSourceRanges: %v", annotationRanges, cidrs)
+		if reflect.DeepEqual(annotationRanges, cidrs) {
```

Comment:
I was wondering why you couldn't just use `reflect.DeepEqual(ic.Status.EndpointPublishingStrategy.LoadBalancer.AllowedSourceRanges, annotationRanges)`; do the types cause a problem (`[]CIDR` versus `[]string`)?

Reply:
I was doing the same at first. That was very strange. It does not complain at compile or run time, but it does not work because of different types (`[]CIDR` versus `[]string`). This cost me a lot of time in the early stage of development as it was difficult to catch.

Reply:
Ah.  That's a good thing to explain in a code comment.  * grin *.